src/Microsoft.TemplateEngine.Cli/New3Command.cs

```diff
@@ -508,6 +522,60 @@ private bool ConfigureLocale()
             return true;
         }
 
+        private bool SyncOptionalWorkloads()
+        {
+            bool isHiveUpdated = false;
+            bool isCustomHive = _commandInput.HasDebuggingFlag("--debug:ephemeral-hive") || _commandInput.HasDebuggingFlag("--debug:custom-hive");
+
+            if (!isCustomHive)
+            {
+                string sdkVersion = EnvironmentSettings.Host.Version;
+
+                try
+                {
+                    List<InstallationRequest> owInstallationRequests = new List<InstallationRequest>();
+                    Dictionary<string, string> owInstalledPkgs = new Dictionary<string, string>();  // packageId -> packageVersion
+                    TemplateLocator optionalWorkloadLocator = new TemplateLocator();
+
+                    IReadOnlyCollection<IOptionalSdkTemplatePackageInfo> owPkgsToSync = optionalWorkloadLocator.GetDotnetSdkTemplatePackages(sdkVersion);
+
+                    foreach (IInstallUnitDescriptor descriptor in _settingsLoader.InstallUnitDescriptorCache.Descriptors.Values)
+                    {
+                        if (descriptor.IsPartOfAnOptionalWorkload)
+                        {
+                            owInstalledPkgs.Add(descriptor.Identifier, descriptor.Details["Version"]);
```

Comment:
- Are we 100% sure that `descriptor.Identifier` is unique here and `Add()` won't throw `ArgumentException` because an entry with the same key already exists?
- Is "Version" guaranteed to exist in `descriptor.Details["Version"]`?

Reply:
1) yes, for the template engine we cannot install the same package twice. so in the cache, every identifier is unique.
2) ok, make sense to add an additional checker.